### PR TITLE
Node Extension GA

### PR DESCRIPTION
**Pull request by duncanpharvey**
### Summary

Updates Node Extension to be GA for apps running in Azure App Service on Windows.

Branched from beta node extension: https://github.com/DataDog/datadog-aas-extension/pull/202

### Changes from Beta

* Removed logic for .NET and Java runtimes since the universal process manager approach will currently only be used for Node
* Build both 32 bit and 64 bit dlls for the Agent Process Manager IIS Module and use the appropriate one depending on the architecture
  * Previously only 32 bit was available
  * dll selection occurs when site extension is installed
* Fixed issue with searching for available UDP ports for DogStatsD
  * Previously checked TCP ports
  * This allows apps running on the same app service plan to all send runtime metrics and custom metrics
* XDT Transform
  * Added `DD_EXTENSION_PATH` environment variable to allow extension to work for both dev and prod
  * Removed `DD_LOG_LEVEL` from XDT Transform since it prevents end users from setting the Datadog log level in Application Settings
  * Preserve options set by customer in `NODE_OPTIONS`
  * Add `node_modules` from site extension to `NODE_PATH` so `dd-trace` is available to send custom metrics and spans
* IIS Module
  * Only listen for `GL_PRE_BEGIN_REQUEST` event. Removed unused code for other events
  * Minor refactor to simplify logic
* Logging
  * Created separate log files depending on the source they were generated from (extension installation, IIS Module, process manager) and updated log pattern to be consistent
  * Fixed issue in process manager that prevented logging to file
* Build script
  * Templates and other files are saved in `src` then move to `content` during the build process
  * Added `content` and other directories created during the build process to `.gitignore`
  * Added script to dynamically set release and development package versions
* Added release pipeline in GitLab
  * Build IIS Module, process manager, and nuget package
  * Push development nuget package to Azure Artifacts and install in dev self monitoring apps before release to prod
  * Create Github Release
  * Push production nuget package to Nuget Gallery and install in prod self monitoring apps

### CODE REVIEW

File: node/AgentProcessManager/AgentProcessManager.vcxproj
Comment (jcstorms1) [inline node/AgentProcessManager/AgentProcessManager.vcxproj:1]: I'm not sure what this is. Do we need it or can it go in .gitignore?

File: node/src/Agent/README.TXT
Comment (jcstorms1) [inline node/src/Agent/README.TXT:5]: Have you confirmed this is still true? I know this used to be the case but I'm not entirely sure that restarts are insufficient.

File: node/src/Agent/dogstatsd.yaml
Comment (jcstorms1) [inline node/src/Agent/dogstatsd.yaml:2]: nice!

File: node/src/applicationHost.xdt
Comment (jcstorms1) [inline node/src/applicationHost.xdt:58]: This is great! Just confirming that the `;.;` isn't a typo. I've never seen that before, but I'm not familiar with Node on Windows.

File: node/process_manager/src/main.rs
Comment (jcstorms1) [inline node/process_manager/src/main.rs:1]: This looks OK to me but I asked Max for a second set of eyes.

File: node/AgentProcessManager/AgentProcessManager.vcxproj
Comment (duncanpharvey) [reply node/AgentProcessManager/AgentProcessManager.vcxproj:1]: Just checked this in a test branch. If this `vcxproj` file is omitted then the `msbuild` commands fail

Screenshot 2024-03-11 at 1 54 14 PM

File: node/src/applicationHost.xdt
Comment (duncanpharvey) [reply node/src/applicationHost.xdt:58]: Confirming that this is **not** a typo. `;` is the delimiter in Windows for multiple paths. This allows dependencies to be resolved in the below order.

1. Customer provided `NODE_PATH`
2. Project directory `.`
3. Site Extension `%XDT_EXTENSIONPATH%\vFOLDERUNKNOWN\Tracer\node_modules`

The goal is to provide `dd-trace` to submit custom metrics and spans without the customer needing to explicitly include `dd-trace` as a dependency in their project.

https://nodejs.org/docs/latest-v18.x/api/path.html#pathdelimiter

**Review by jcstorms1 — APPROVED** [review]
Great work!
Let's wait for the review from Max and reach out to .NET team. We don't want to mess with their pipeline.

File: node/AgentProcessManager/AgentProcessManager.vcxproj
Comment (lucaspimentel) [reply node/AgentProcessManager/AgentProcessManager.vcxproj:1]: This is the Visual Studio project file for C++ projects and is required to build with `msbuild`. It is analogous to the `csproj` file for C# (CSharp) projects or `vbproj` for Visual Basic.

File: .gitignore
Comment (lucaspimentel) [inline .gitignore:23]: It's always ok to ignore all `.user` files in VS.

```suggestion
*.user
```

**Review by andrewlock — APPROVED** [review]
From the .NET side, AFACIT the package produced still looks correct to me.

Only thing that stood out is it's using an _old_ package version (2.46.0) but I assume that's just because you're not rebased on master? 

I don't _think_ this impacts the dev build for .NET at all, so shouldn't be a problem there

File: node/process_manager/src/main.rs
Comment (maxday) [inline node/process_manager/src/main.rs:22]: what do you think about extracting this logic into a pure function? this way you could add a unit tests against it as it seems that this logic is pretty isolated from the remaining of the main function

**Comment by duncanpharvey** [comment]
Excellent! Yeah this is branched off an older version of master so it doesn't have some of the recent commits updating the .NET Extension version

File: node/process_manager/src/main.rs
Comment (maxday) [inline node/process_manager/src/main.rs:86]: what about returning a boolean from this function? I feel it might be a bit dangerous be be able to call `spawn_helper` without knowing whether it has failed or nor?

File: node/process_manager/src/main.rs
Comment (duncanpharvey) [reply node/process_manager/src/main.rs]: Agreed, changed the return type to void

File: node/process_manager/src/main.rs
Comment (duncanpharvey) [reply node/process_manager/src/main.rs]: Agreed, done!

File: node/process_manager/src/main.rs
Comment (duncanpharvey) [reply node/process_manager/src/main.rs]: I updated `spawn_helper` to return a bool with an error message and am now logging the errors from outside the function.

File: node/process_manager/src/main.rs
Comment (maxday) [inline node/process_manager/src/main.rs:78]: `map_err` might help you there: 
```
env::var("DD_API_KEY").map_err(|_| "DD_API_KEY not provided".to_string())?;
```

**Review by maxday — APPROVED** [review]

File: node/process_manager/src/main.rs
Comment (duncanpharvey) [reply node/process_manager/src/main.rs:1]: Great idea! I'll work on the tests and add them to a future release

File: node/process_manager/src/main.rs
Comment (duncanpharvey) [reply node/process_manager/src/main.rs]: Just committed this change

**Review by jcstorms1 — APPROVED** [review]
Great work!

**Pull request merged**
